File: Proxy02.py
# Objeto Real

class Produto:
    def __init__(self, nome, preco):
        self.nome = nome
        self.preco = preco

    def exibir_detalhes(self):
        print(f"Produto: {self.nome}, Preço: R$ {self.preco:.2f}")

# Proxy - CLASSE PROXY PRODUTO

class ProxyProduto:
    def __init__(self, produto):
        self.produto = produto

    def exibir_detalhes(self):
        print("Aguardando autenticação...")
        self.autenticar()
        self.produto.exibir_detalhes()

    def autenticar(self):
        print("Autenticando usuário...")
        # INSERIR LOGICA DE AUTENTICACAO AQUI
        print("Usuário autenticado com sucesso")

# Uso do proxy cliente

produto_real = Produto("Camiseta", 39.90)
proxy_produto = ProxyProduto(produto_real)

proxy_produto.exibir_detalhes()

File: Proxy02.js
// Objeto Real

class Produto{
    constructor(nome, preco){
        this.nome = nome;
        this.preco = preco;
    }

    exibirDetalhes(){
        console.log(`Produto: ${this.nome}, Preço: R$ ${this.preco.toFixed(2)}`);
    }
}

// Proxy -  CLASSE PROXY PRODUTO

class ProxyProduto{
    constructor(produto){
        this.produto = produto;
    }

    exibirDetalhes(){
        console.log("Aguardando autenticacao...");
        this.autenticar();
        this.produto.exibirDetalhes();
    }

    autenticar(){
        console.log("Autenticando usuario...");
        // INSERIR LOGICA DE AUTENTICACAO AQUI
        console.log("Usuário autenticado com sucesso");
    }
}

// Uso do proxy cliente

const produtoReal = new Produto("Camiseta", 39.90);
const proxyProduto = new ProxyProduto(produtoReal);

proxyProduto.exibirDetalhes();